### DISCUSSION TITLE
3D Local thickness vs 2D slice-wise local thickness

### DISCUSSION THREAD

**Discussion by OhmFF** [PMEAL/porespy#657] [Q&A] [answered]
Hello all, I am analyzing pore size distribution from FIB-SEM 3D volume. I found that porespy's `local_thickness` is the most promising tool out there. However, I have the following questions.
**1.) The results from 3D local thickness and 3D from stacked 2D-slice-wise local thickness are different**

![2022-03-15 22_10_54-PowerPoint Slide Show  -  Presentation1](https://user-images.githubusercontent.com/42203511/158384936-7f81d7d2-1345-4623-ab42-53af2f3af9f7.png)
As you can see, the pore size distribution of the two approaches in the same image is totally different. 2D local thickness does not have 1-voxel-size artifacts around the edge of the pore. Larger "circles" can fill the pore space because the algorithm takes 1 slice by 1 slice, resulting in larger pore size distribution. Whereas, when applying local thickness in 3D pore space, it has the influence of 3D pore space due to segmentation or whatsoever, the "sphere" cannot be fully inscribed the pore space and produce many 1-voxel-size pores. From my understanding, when analyzing a 3D volume, it should be the 3D local thickness rather than a stack from 2D-local thickness because we really put "spheres" not "circles (or disks)" in pore space. Is it correct?

How can we minimize this 1-voxel-size pore..? Is there anything we can do about it? Because of this problem, the interpretation of pore size distribution from local thickness is skeptical for me. 

**2.) How do you consider the `sizes` parameter of local_thickness?**
Due to the 1-voxel-size pore artifact, I have come across these discussions. ((https://github.com/PMEAL/porespy/issues/197)) ((https://github.com/PMEAL/porespy/issues/64))
It seems like "sizes" to invade is also important. How do you know how much should it be (default is 25)? Is it from the maximum value of `np.unique(spim.distance_transform_edt(image))`? 

I have tried with `sizes = 100` with `mode = 'hybrid'` and `mode = 'dt'`. But it does not make any difference.. here is the result.
![2022-03-15 22_21_41-Presentation1 - PowerPoint](https://user-images.githubusercontent.com/42203511/158390365-ac07cb49-2021-48a3-9eae-98c04b70d118.png)

**3.) What is the best parameter setting (`sizes`, `mode`, and `divs`) for `local_thickness` to analyze `pore_size_distribution` in a 3D image**
There is one parameter that I don't understand which is `divs`. When analyzing a 3D image, `divs = 2` is recommended??
I tried experimenting with it, but the results are the same if `sizes ` are large enough (i.e., 100). The current setting that I am using is `local_thickness(image, sizes = 100, mode = 'hybrid', and divs = 1)`. What could be the best parameter setting (if there is)?

Thank you very much. I really appreciate the work of PMEAL to make such a good tool and discussion forum!

**Comment by jgostick** [comment, marked as answer]
Hi @OhmFF 
Thanks for providing so much detail with your question, it really helps.  Before I answer, I will point out that we are in the middle of a major overhaul of the documentation, particularly the examples (see [this PR](https://github.com/PMEAL/porespy/pull/630)).  The new examples aim to specifically outline what each argument does, so that would answer several of your questions.  But let me answer them below:

1.  As you say, 2D vs 3D is going to give different sizes since a circle can fit into a 2D region, but a similar sized sphere almost certainly will not fit once the "third dimension" is included, which contains solid obstacles in the other planes.  So the trend you are seeing is correct, with the 2D sizes being shifted to larger values.  As for the proliferation of 1's in the image this is probably also due to 3D effects.  Basically the sphere cannot reach into the corners.  It's probably a valuable signal about the structure of your material.  I don't think it's a bug because this filter gets used a lot and has proven to be pretty robust, but anything is possible.
2. ``sizes`` controls the sizes of spheres to use.  If you set ``sizes=5`` the image will come back with only 5 sizes in it.  The default is 25.  You can also specify your own and pass in as an array, like ``sizes=[25, 22, 19, 15, 12, 10, 8, 6, 5, 4, 3]``.  Although, this just means spheres of size 2 and 1 will not be drawn, so those pore spaces will remain empty.  You could just remove the small values with ``lt = lt*(lt > 3)``.  Of if you want all the small spaces to be assigned to 3 instead of 1, add ``lt[lt < 3] = 3`` and ``lt = lt*im`` (to remove the 3's added to the solid phase locations).
3. The mode argument should have no effect on the result, only the performance.  ``divs`` is for parallellization, with 1 meaning no parallelization, and >1 meaning the image is broken in the chunks and processed on each core.  Again, this should have not effect on the results.

**Reply by OhmFF** [reply]
@jgostick Thank you very much for the fruitful discussion, sir. I am a big fan of porespy and looking forward to seeing the new documentation!

1.) I don't think it's a bug as well, it is due to 3D effects as you said. In this case, pore size distribution from which process is more realistic, which one should I trust..? (Sorry, this might sound stupid. My focus is on several nanometer sizes, and I planned to combine multiple experimental techniques, so accurate pore size distribution is very important.)
2.) Understood, it helps me understand the `sizes` now.
3.) Interesting, in that case, does it mean that `divs = 2` will give a faster calculation because of the parallelization?

**Reply by jgostick** [reply]
1. Instead of local thickness you could try chords, which are 'directional' so they may not have the problem of fitting into the small corners.  You can get chords in 3 perpendicular directions to detect anisotropy, which is my favorite features.  You could also do the radial density function, which is basically a histogram of the distance transform.  This has lots of 1s but this is normal part of the analysis.
2. Good!
3. If you have lots of RAM and multiple cores, then divs=2 or 3 will chunk up the image into 2x2x2 or 3x3x3 blocks (with some overlap) and process in parallel.  This uses dask under the hood.  Try using ``%timeit ps.filters.local_thickness(im, divs=2 <or 1 or 3>)`` to see if it helps.

**Reply by OhmFF** [reply]
Thank you for your prompt suggestions. I will try that and share the results !
